### PR TITLE
Optimizer State and Master Weight Offloading

### PR DESCRIPTION
**Pull request by hxbai**
# What does this PR do ?
dev branch: #2760 

### Summary

This PR introduces optimizer state offloading to CPU for the `DistributedOptimizer`, enabling significant GPU memory savings during training by temporarily moving optimizer states (exp_avg, exp_avg_sq) and master weights to CPU memory when not in use.

### Motivation

During the forward and backward passes, optimizer states occupy GPU memory but are not actively used. For large models, these states can consume a substantial portion of GPU memory. By offloading optimizer states to CPU after `optimizer.step()` and reloading them before the next step, we can reclaim this GPU memory for other operations like activation checkpointing or larger batch sizes.

### Performance & Memory Savings

**Memory Savings:**
- On **DeepSeek-V3**, this feature saves **15-20GB** of GPU memory (with 0.1-0.2s/iter overhead on GB200)

**Comparison with Optimizer CPU Offloading:**

| Aspect | CPU Offloading (ZeRO-Offload style) | State Offloading (This PR) |
|--------|-------------------------------------|----------------------------|
| **Where optimizer runs** | CPU | GPU |
| **D2H/H2D frequency** | Every step (gradients + params) | Every step (states only) |
| **Compute location** | Adam step on CPU | Adam step on GPU |
| **Best for** | Memory-constrained, bandwidth-limited | High-bandwidth interconnects (NVLink, GB200) |

More details:
- With higher H2D/D2H bandwidth such as **GB200**, the state offloading has significantly less overhead
- Async transfer overlapping
- Optimizer step still runs on GPU, avoiding CPU compute bottleneck
- Pinned memory enables maximum PCIe/NVLink bandwidth utilization
- Currently requires TE FusedAdam optimizer

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @megatron-oncall.

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

Feel free to message or comment the @megatron-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

**Comment by Phlip79** [comment]
**We are changing our review process and marking all open, unlabeled PRs as draft. This change will go in effect starting once https://github.com/NVIDIA/Megatron-LM/pull/3659 is merged.**

Moving forward, all PRs will be required to start as draft PRs. If you wish to get your PR merged, mark your PR as “Ready for review”. Read more about the new process at [submit.md](https://github.com/Phlip79/Megatron-LM/blob/d3763629bcdd04acca52477af7eb8795090426ff/docs/developer/submit.md).

File: megatron/training/arguments.py [hunk 1]
```diff
@@ -1274,13 +1217,18 @@ def validate_args(args, defaults={}):
             "must be used in conjunction with `--fp8-recipe delayed`."
         )
 
+    if args.offload_optimizer_states:
```
Comment (ericharper) [inline megatron/training/arguments.py:1522]: does this work with `args.optimizer_cpu_offload` ? Can we or should we ever offload both?

Comment (hxbai) [reply megatron/training/arguments.py:1522]: They cannot work together. optimizer_cpu_offload offloads more things, including states, grads, etc., and does the optimizer computation on CPU. So, there is no need to use them together. But CPU computation is slow, this feature just copies states from/to GPUs and still has the computation on the GPU, which is much faster on GB200-like systems.

Comment (ericharper) [reply megatron/training/arguments.py:1522]: do we need to add an assert for this then?

File: megatron/core/optimizer/cpu_offloading/optimizer_state_offloader.py [hunk 2]
```diff
@@ -0,0 +1,315 @@
+# Copyright (c) 2025, NVIDIA CORPORATION. All rights reserved.
```
Comment (jaredcasper) [inline megatron/core/optimizer/cpu_offloading/optimizer_state_offloader.py]: Wrong year.

**Review by Phlip79 — CHANGES_REQUESTED** [review]
Can you please update the documentation at `megatron/core/optimizer/cpu_offloading/README.md`?

File: megatron/core/optimizer/cpu_offloading/optimizer_state_offloader.py [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+
+import torch
+
+if TYPE_CHECKING:
+    from megatron.core.optimizer.distrib_optimizer import DistributedOptimizer
+
+
+class OptimizerStateOffloader:
+    """
+    Manages offloading of optimizer states and master weights to CPU.
```
Comment (jaredcasper) [inline megatron/core/optimizer/cpu_offloading/optimizer_state_offloader.py]: Can we not use the term "master weights" throughout? "Primary weights" or "fp32 weights" instead?

File: tests/unit_tests/test_optimizer_state_offloading.py [hunk 4]
```diff
@@ -0,0 +1,337 @@
+# Copyright (c) 2025, NVIDIA CORPORATION. All rights reserved.
```
Comment (jaredcasper) [inline tests/unit_tests/test_optimizer_state_offloading.py]: Wrong year.

File: megatron/training/arguments.py (hunk 1 above)
Comment (hxbai) [reply megatron/training/arguments.py:1522]: Added.

File: megatron/core/optimizer/cpu_offloading/optimizer_state_offloader.py (hunk 2 above)
Comment (hxbai) [reply megatron/core/optimizer/cpu_offloading/optimizer_state_offloader.py]: fixed

File: tests/unit_tests/test_optimizer_state_offloading.py (hunk 4 above)
Comment (hxbai) [reply tests/unit_tests/test_optimizer_state_offloading.py]: fixed

File: megatron/core/optimizer/cpu_offloading/optimizer_state_offloader.py (hunk 3 above)
Comment (hxbai) [reply megatron/core/optimizer/cpu_offloading/optimizer_state_offloader.py]: Other places, including the TE flag and attributes, use "master weights". It is confusing unless we rename all the use cases from TE to MCore.

**Review by Phlip79 — APPROVED** [review]

File: megatron/core/optimizer/cpu_offloading/optimizer_state_offloader.py (hunk 3 above)
Comment (jaredcasper) [reply megatron/core/optimizer/cpu_offloading/optimizer_state_offloader.py]: I can't control what TE does, but I have consistently been trying to remove "master" terminology from mcore. This is a company policy: https://nvidia.atlassian.net/wiki/spaces/ReLingo/overview. Last I checked the only use of this in mcore was when necessary to interact with TE, and that does not seem to be the case here.

Besides company policy it's a bad term, it's not clear what it means. Rename it to be more descriptive.

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/2811/commits/ae4db531a658ae49398bc08eeb11b348e169459d

File: megatron/core/optimizer/cpu_offloading/optimizer_state_offloader.py (hunk 3 above)
Comment (hxbai) [reply megatron/core/optimizer/cpu_offloading/optimizer_state_offloader.py]: Done.
